File: leetcode/p1680.cpp
class Solution {
public:
    int concatenatedBinary(int n) {
        int M = 1000000007;
        unsigned long long int total=1;
        int i=2, b=1;
        while (i<=n) {
            b++;
            int t = 1 << b;
            //cout << b << endl;
            for (;i<t;i++) {
                total = (total << b) + i;
                //cout << total << endl;
                if (total >= M) total = total%M;
                if (i == n) return total;
            }
        }
        return total;
    }
};

File: leetcode/p1680.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        k = 10**9 + 7
        mod_k = 1
        total = 0
        for i in range(n, 0, -1):
            t = mod_k * i
            total += mod_k * i
            if total > k:
                total -= k
            #print(t, mod_k)
            mod_k = (mod_k * 2**(len(bin(i))-2)) % k
        return total % k
